File: m4d/ClientApp/src/models/DanceObject.ts
import { jsonMember, jsonObject } from "typedjson";
import { Meter } from "./Meter";
import { NamedObject } from "./NamedObject";
import { TempoRange } from "./TempoRange";

@jsonObject
export class DanceObject extends NamedObject {
  @jsonMember(Meter) public meter!: Meter;
  @jsonMember(TempoRange) public tempoRange!: TempoRange;
  @jsonMember(String) public blogTag?: string;

  public get baseId(): string {
    return this.id.substr(0, 3);
  }
}

File: DanceLib/DanceObject.cs
using System.Collections.Generic;
using Newtonsoft.Json;

namespace DanceLibrary;

public class DanceObject
{
    public virtual string Id { get; set; }

    public virtual string Name { get; set; }

    public virtual Meter Meter { get; set; }

    public virtual TempoRange TempoRange { get; set; }

    public virtual string BlogTag { get; set; }

    public virtual List<string> Synonyms { get; set; }

    public virtual List<string> Searchonyms { get; set; }

    [JsonIgnore]
    public string CleanName => SeoFriendly(Name);

    public static string SeoFriendly(string name)
    {
        return string.IsNullOrWhiteSpace(name) ? name : name.Replace(' ', '-').ToLower();
    }
}
